### DISCUSSION TITLE
Flash ESP-S3 how to...

### DISCUSSION THREAD

**Discussion by tombloemker** [toverainc/willow#158] [General]
I'm missing something in the read me file. 
 
SYSTEM DEPENDANCIES 
Install Tio instruction mentions Mac, Linux but not windows
I followed instructions on the Tio page which is installed via MSYS2
I can list devices but no ID is listed
dev/tty0
dev/tty3
(unplugging the ESP makes tty3 disappear)

the following code git clone https://github.com/toverainc/willow.git && cd willow 
result -bash: git: command not found

I assume I need to build a container via docker? No link...  

Can someone add clarification?

==============================================
I woke up the next morning thinking...
"Hey, are these instructions meant to install tio onto my Home Assistant Pi and flash from there?"

**Comment by nikito** [comment]
From the above it looks like git isn't installed? Most of these instructions are for Linux or Mac, not sure how well Windows itself would work. Might be easier to either try WLS (windows Linux subsystem) or maybe spin up a Linux vm? 🙂

**Reply by tombloemker** [reply]
Ugh, okay. I really appreciate the direction. If WLS has good documentation I will go in that route. Any Youtube cliff notes or good reading recommendations for this novice?

**Reply by hamishcunningham** [reply]
I've not tried WSL, but I think that it, or a Linux VM, would be an easier route than trying to install via one of the emulators or compatibility layers (e.g. MSYS) running in native Windows.

Essentially Willow and WIS are in the process of rapidly transforming from a developer-only type of build process to a more end-user friendly type, but this is a) a lot of work and b) not so easy in the case of Windows :(

You might take the plunge and learn Linux, the only operating system you'll ever need. I've been running essentially same tools these last 20 years, and they work better than ever :)

**Reply by tombloemker** [reply]
Okay, WSL installed, the git command ran but now I'm stuck on the next step <./utils.sh build-docker>  I assume I'm missing part of the syntax?

**Reply by hamishcunningham** [reply]
What error?

**Reply by tombloemker** [reply]
Sorry, message below
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/build?buildargs=%7B%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile&labels=%7B%7D&memory=0&memswap=0&networkmode=default&rm=1&shmsize=0&t=willow%3Alatest&target=&ulimits=null&version=1": dial unix /var/run/docker.sock: connect: permission denied

**Reply by tombloemker** [reply]
I figured it out, I change the step to <sudo ./utils.sh build-docker>

**Comment by hamishcunningham** [comment]
Yes, or add yourself to the docker group slightly preferred :)

**Comment by kimocal** [comment]
I created a Ubuntu 23 VM and got to the `./utils.sh install` step but it gives me this message:  `You already have a deps directory - exiting` 

Any suggestions? 

**EDIT**  I rm -rf the willow directory and redid the steps.  Install step working now.

**Reply by DustyMustard** [reply]
I tried in WSL2 and gave up at the part where you try to find the com port.

If I were you I would install Linux on a flash drive abd do it that way. Install on Windows needs prior experience.

**Comment by hamishcunningham** [comment]
looks like you've already done the install step previously?
if you want to do it again from scratch try
rm -rf deps
hth
h

On Fri, 16 Jun 2023 at 21:31, kimocal ***@***.***> wrote:

> I created a Ubuntu 23 VM and got to the ./utils.sh install step but it
> gives me this message: You already have a deps directory - exiting
>
> Any suggestions?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/toverainc/willow/discussions/158#discussioncomment-6200927>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBRBFHHDJQ34OHFQKHBNW3XLS7AHANCNFSM6AAAAAAZABTQ5I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

*Hamish Cunningham*
Professor of Computer Science, University of Sheffield, UK


*I used to hope that as time passed I would get older and wiser,
        but it seems that in fact I just get odder and wider.*

*Phone: 07740 149 842 / 0114 222 1891             * *unphone.net*
<https://unphone.net/>
***@***.***            ***@***.***
https://hamish.gate.ac.uk    *https://iot.unphone.net
<https://iot.unphone.net/>*
https://gate.ac.uk                 https://pi.gate.ac.uk

*https://www.sheffield.ac.uk/dcs/people/academic/hcunningham
<https://www.sheffield.ac.uk/dcs/people/academic/hcunningham>*

**Comment by hamishcunningham** [comment]
It is very difficult / impossible to get WSL2 to pass through the com port.
(It was possible in WSL1...)
So a VM that will pass the port through, or copying the .bin and flashing
from Windows are better options.

On Sat, 17 Jun 2023 at 09:23, DustyMustard ***@***.***> wrote:

> I tried in WSL2 and gave up at the part where you try to find the com port.
>
> If I were you I would install Linux on a flash drive abd do it that way.
> Install on Windows needs prior experience.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/toverainc/willow/discussions/158#discussioncomment-6203245>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBRBFBF6PBQ5J45VBGXHHLXLVSOPANCNFSM6AAAAAAZABTQ5I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

*Hamish Cunningham*
Professor of Computer Science, University of Sheffield, UK


*I used to hope that as time passed I would get older and wiser,
        but it seems that in fact I just get odder and wider.*

*Phone: 07740 149 842 / 0114 222 1891             * *unphone.net*
<https://unphone.net/>
***@***.***            ***@***.***
https://hamish.gate.ac.uk    *https://iot.unphone.net
<https://iot.unphone.net/>*
https://gate.ac.uk                 https://pi.gate.ac.uk

*https://www.sheffield.ac.uk/dcs/people/academic/hcunningham
<https://www.sheffield.ac.uk/dcs/people/academic/hcunningham>*

**Comment by nikito** [comment]
Another option people have been doing is to use the dist build, then flash using the esphome web tool. May be worth looking into for your use case? I think most of these issues are going to go away though when release 1.0 comes as it will support flashing similar to esphome via web browser. 🙂

**Comment by kimocal** [comment]
I tried going the WSL route and hit too many roadblocks.  Ended up spinning up a Ubuntu 23 VM in unRaid and got to the flashing part but had some USB communications issues.  Then [I found a comment in the Issues section](https://github.com/toverainc/willow/issues/60#issuecomment-1553818139) and used the willow-dist.bin and flashed it on my Win 10 laptop.  Lots of good info to help get people through the process.  Its helping me learn too.

**Reply by kristiankielhofner** [reply]
Glad you got it figured out!

When first releasing Willow I assumed most issues would come from what I considered to be the "hard part" - wake, speech recognition, audio generally.

It turns out that flashing is by FAR the biggest source of issues. In the next few weeks we will be releasing Willow 1.0 which will support easy flashing of pre-built binaries from a web browser, dynamic configuration, over the air updates, etc.

Hopefully we'll be able to put these bothersome flash issues behind us soon!

**Comment by tombloemker** [comment]
My hope is to answer my original question for anyone who can't wait for the July release.  (Don't rely on this alone)
I did install Ubuntu and WSL per the instructions (I think WSL is initiated from powershell?)
I did Tio but may have had permission issues CMD: sudo gpasswd -a $USER docker  OR CMD: sudo usermod -aG docker $USER (To add yourself to the docker, I'm not really sure where in the process I did this)
CMD: apt-get install tio  
CMD: git clone https://github.com/toverainc/willow.git && cd willow
CMD: ./utils.sh build-docker   
CMD: ./utils.sh docker       (Comment: You don't build a container the command does it for you)
CMD: ./utils.sh install
CMD: ./utils.sh config       (If you have to change your ESP32 configuration, this starts the configuration program)
CMD: ./utils.sh build
CMD: ./utils.sh dist           (This creates the willow-dist.bin file)
My path to the file was  \\wsl.localhost\Ubuntu\mnt\wslg\distro\home\tomb\willow\serve

[ESPTool](https://espressif.github.io/esptool-js/)...  It worked the first time I used it, something happened when I re-flashed. 
I got an error "no file selected for row 1!"  I deleted my link and cleared my cache and went back to the [original link](https://espressif.github.io/esptool-js/).  This fixed my problem.  Select 115200 Baud press Connect; A Serial Port window will pop up, if you don't know which PORT is your ESP32, unplug it and it will disappear, reconnect and select; Press Erase Flash;  Change Flash address to 0x0; Press Choose file, Find and select Willow-dist.bin file; Press Program (Let it run until it says "Leaving..." then Press disconnect.  

Your device will reboot... 
I used [Termite](https://www.compuphase.com/software_termite.htm) to monitor the output of the device.
I had a terrible time understanding what I was looking at, my homeassistant.local address wouldn't resolve. I ended up using the IP of HA hub and that worked. I tried one command over  and over, in termite I noticed verbal commands produced much code, one line looks like
WILLOW: home assistant response_type: action_**done** or  WILLOW: home assistant response_type: **error**
Then I tried a different HA device and it worked! 
In home assistant select, settings, devices & services, entites, click on an entity;click settings gear icon in upper right corner; click Voice assistants;  Add extra Aliases based on the text resolved on the willow device.    

notes: making configuration changes required re-initiating the configuration menu. 
Open Ubuntu
cd willow
Use the CMD's above starting with ./utils.sh docker  (I might be wrong but it worked for me)

**Reply by Takore2002** [reply]
I had spent a week trying to get everything to work in either Powershell or Ubuntu WSL but ended up having to do a mix of the two.

After running ./utils.sh build in Ubuntu I had to switch to Powershell run
wsl --set-default Ubuntu //sets the default linux distro)
usbipd wsl list //find the BUSID for the device)
usbipd wsl attach -b (BUSID) // to share the port with Ubuntu

After that I was able to follow the rest of the flashing process in the ReadMe.

**Reply by tombloemker** [reply]
I wish I would have thought of that... 
After testing... I assume you are running Win11? 
No Joy on Win10Pro.  :(   So far...

**Reply by tombloemker** [reply]
Still trying to see if there is a solution here...
[These instructions have changed.
Updating to usbipd-win version 2.0.0 or higher will require these new instructions, even if the old instructions were followed before.](https://github.com/dorssel/usbipd-win/wiki/WSL-support)

**Reply by Takore2002** [reply]
Sorry didn't notice your first response. I am actually on Windows 10pro as well. It was a while ago that I did it and I vaguely remember needing to install a couple packages in Powershell before I could get those commands to work. Which part is it that you are running into issues on?

Edit: looking through the command history in powershell I think most of what I remember trying to install was everything needed to replace Ubuntu and didn't have any luck with.

**Reply by tombloemker** [reply]
In the link above they say run these commands in ubuntu
From within WSL, install the user space tools for USB/IP and a database of USB hardware identifiers. On Ubuntu 20.04 LTS, run these commands:

sudo apt install linux-tools-virtual hwdata
sudo update-alternatives --install /usr/local/bin/usbip usbip `ls /usr/lib/linux-tools/*/usbip | tail -n1` 20

I got the first command to run but but not the second, I don't know anything about "user space tools for USB/IP and a database of USB hardware identifiers."

Do you have better instructions to upgrade powershell?

**Reply by Takore2002** [reply]
Ah, gotcha. I was never able to get WSL working in Ubuntu on Windows 10 which is why I had to use Powershell as well.

After using usbipd wsl attach -b (BUSID) in Powershell to attach the port to the Ubuntu instance I was able to use the tio commands in Ubuntu to see and interact with the port.

**Comment by hamishcunningham** [comment]
Great, well done!

**Comment by Cjkeenan** [comment]
I am using https://github.com/ublue-os specifically the main cinnamon image and I managed to install the docker image and get in, but on launch it gives these errors:
```
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:

  idf.py build

Traceback (most recent call last):
  File "/opt/esp/idf/tools/idf.py", line 821, in <module>
    main()
  File "/opt/esp/idf/tools/idf.py", line 737, in main
    cli = init_cli(verbose_output=checks_output)
  File "/opt/esp/idf/tools/idf.py", line 661, in init_cli
    project_dir = parse_project_dir(standalone_mode=False, complete_var='_IDF.PY_COMPLETE_NOT_EXISTING')
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 1054, in main
    with self.make_context(prog_name, args, **extra) as ctx:
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 920, in make_context
    self.parse_args(ctx, args)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 1378, in parse_args
    value, args = param.handle_parse_result(ctx, opts, args)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 2360, in handle_parse_result
    value = self.process_value(ctx, value)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 2316, in process_value
    value = self.type_cast_value(ctx, value)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/core.py", line 2304, in type_cast_value
    return convert(value)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/types.py", line 82, in __call__
    return self.convert(value, param, ctx)
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/types.py", line 886, in convert
    self.fail(
  File "/opt/esp/python_env/idf4.4_py3.8_env/lib/python3.8/site-packages/click/types.py", line 135, in fail
    raise BadParameter(message, ctx=ctx, param=param)
click.exceptions.BadParameter: Path '/willow' is not readable.
```
It seems the permissions of `/willow` are set to `build:build`:
```
root@e6dda8a68ffb:/willow# cd ..
root@e6dda8a68ffb:/# ll
total 24
drwxr-xr-x.   1 root  root   186 Jul 10 04:00 ./
drwxr-xr-x.   1 root  root   186 Jul 10 04:00 ../
-rwxr-xr-x.   1 root  root     0 Jul 10 04:00 .dockerenv*
lrwxrwxrwx.   1 root  root     7 Apr 12 02:03 bin -> usr/bin/
drwxr-xr-x.   1 root  root     0 Apr 15  2020 boot/
drwxr-xr-x.   5 root  root   360 Jul 10 04:00 dev/
drwxr-xr-x.   1 root  root  1732 Jul 10 04:00 etc/
drwxr-xr-x.   1 root  root    10 Jul 10 03:56 home/
lrwxrwxrwx.   1 root  root     7 Apr 12 02:03 lib -> usr/lib/
lrwxrwxrwx.   1 root  root     9 Apr 12 02:03 lib32 -> usr/lib32/
lrwxrwxrwx.   1 root  root     9 Apr 12 02:03 lib64 -> usr/lib64/
lrwxrwxrwx.   1 root  root    10 Apr 12 02:03 libx32 -> usr/libx32/
drwxr-xr-x.   1 root  root     0 Apr 12 02:03 media/
drwxr-xr-x.   1 root  root     0 Apr 12 02:03 mnt/
drwxr-xr-x.   1 root  root     6 Jun  8 05:32 opt/
dr-xr-xr-x. 348 root  root     0 Jul 10 04:00 proc/
drwx------.   1 root  root    74 Jul 10 03:56 root/
drwxr-xr-x.   1 root  root    40 Apr 12 02:06 run/
lrwxrwxrwx.   1 root  root     8 Apr 12 02:03 sbin -> usr/sbin/
drwxr-xr-x.   1 root  root     0 Apr 12 02:03 srv/
dr-xr-xr-x.  13 root  root     0 Jul 10 03:51 sys/
drwxrwxrwt.   1 root  root   110 Jun  8 05:35 tmp/
drwxr-xr-x.   1 root  root   102 Apr 12 02:03 usr/
drwxr-xr-x.   1 root  root    90 Apr 12 02:06 var/
drwxr-xr-x.   1 build build  534 Jul 10 03:36 willow/
root@e6dda8a68ffb:/# ll /willow
ls: cannot open directory '/willow': Permission denied
```
Also when I su to build it does not seem to help, also `sudo chown -R root:root /willow` says permission denied. I am under the assumption it is something to do with my "cutting edge" operating system, but I am at a loss for now.

Any ideas would be greatly appreciated.
